core/router/gateway/gateway.go

```diff
@@ -51,18 +50,17 @@ func (g *Gateway) SetAuth(token string, authenticated bool) {
 	g.mu.Lock()
 	defer g.mu.Unlock()
 	g.authenticated = authenticated
-	if g.MonitorStream != nil {
-		if token == g.token {
-			return
-		}
-		g.Ctx.Debug("Stopping Monitor stream (token changed)")
-		g.MonitorStream.Close()
+	if token == g.token {
+		return
 	}
 	g.token = token
-	if g.Monitor != nil {
-		g.Ctx.Debug("Starting Gateway Monitor Stream")
-		g.MonitorStream = g.Monitor.NewGatewayStreams(g.ID, g.token)
-	}
+	g.MonitorStream.Reset()
+}
+
+func (g *Gateway) GetToken() string {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
```

Comment:
This function is short enough for `defer` to produce significant overhead.
https://lk4d4.darth.io/posts/defer/

```
g.mu.RLock()
token := g.token
g.mu.RUnlock()
return token
```